Makefile.am

```diff
@@ -43,20 +43,33 @@ EXPECTS = $(wildcard tests/*.expected)
 .PHONY: update_expected
 update_expected: $(EXPECTS)
 
-tests/%.expected: tests/%.sil force | sile
+tests/%.expected: tests/%.sil force | sile testprep
 	$(LOCALTESTFONTS) ./sile -b debug $< -o $@
 
 ACTUALS = $(addsuffix .actual,$(basename $(wildcard tests/*.expected)))
 
 .PHONY: update_actuals
 update_actuals: $(ACTUALS)
 
-tests/%.actual: tests/%.sil | sile
+tests/%.actual: tests/%.sil | sile testprep
 	$(LOCALTESTFONTS) ./sile -b debug $< -o $@
 
 .PHONY: force
 force: ;
 
+.PRECIOUS: .fonts/%
+
+TESTFONTFILES  = NotoSansCJK-Regular.ttc
```

Comment:
I don't know where you're going with this long term but for me the obvious way forward here is to have a CSV/YAML/JSON/whatever file in the `tests/` directory which lists all the font files required and the URLs to download them from.

Reply:
It's not quite that simple. If all we needed was a map of files and URLs that would be one thing, but it isn't a straight map like that. Different sites have different download restrictions, some are in archives, some not. Doing this with `make` recepies allows us to do some pattern matching and script the download/extraction process differently for each font or group of fonts by pattern.

I did look at extracting a list of fonts needed using SILE's debug output itself and will look at doing it again as a last step here, but decided listing them by hand at least during the transition while I went through each test trying to replicate issues was a lot easier. Once they all work this way it should be easy to compare if the auto generated list of required fonts is actually accurate.

Reply:
OK, before we go much further on this I wonder if the prerequisite is to try to reduce the number of unique fonts used in tests. Currently it looks like this:

| Font name | Number of tests using | Tests using
| ---------- | ---- | --- |
| 	 | 1 | color-fonts* |
| Amiri | 1 | bug-192-full |
| Amiri Quran | 1 | sura-2 |
| Crimson | 1 | bug-307* |
| DejaVu Sans | 1 | bug-76 |
| DejaVu Sans Mono | 2 | bug-132,bug-317 |
| Gentium | 4 | 049EPH,bug-124_upstream,chapterverse,combining |
| Gentium Book Basic | 1 | nondeterminism |
| Gentium Plus | 81 | |
| Kohinoor Devanagari | 1 | bug-75_darwin |
| Lateef | 1 | arabic-scripts |
| Libertinus Serif | 1 | bug-307* |
| Linux Biolinum O | 1 | balanced* |
| Linux Libertine O | 4 | balanced*,bug-283,bug-79,linespacing |
| Montserrat | 1 | bug-307* |
| Noto Naskh Arabic | 4 | bug-124_upstream,bug-192,bug-76,mini-arabic |
| Noto Sans CJK JP | 3 | bug-524,bug-525,vertical |
| Noto Sans Kannada | 1 | kannada |
| Noto Sans Malayalam | 1 | bug-308 |
| SBL Hebrew | 1 | bidi |
| Scheherazade | 2 | bug-124_upstream,bug-192-full |

(* marks knownbad)

It would be nice to see if we can tickle the same bugs without some of these fonts or by substituting other ones.

Reply:
Some comments on these:

* `arabic-scripts`, `bug-192-full`: Any Arabic would do.
* `bug-124_upstream`: Is these even being tested? Should it be? It's fixed now. So Scheherazade can go.
* `bug-76`: DejaVu Sans can be replaced by any Latin.
* `bug-132`, `bug-317`: Actually any secondary font would do; doesn't even need to be monospaced.
* Gentium only appears where expectations files are old.
* `nondeterminism`: I don't even remember what this is meant to test. I don't think it's valid any more.

Reply:
I've been reducing the number of fonts as I go along by trying to reproduce bugs with one of the fonts already required before requiring a new one for the test. I suspect almost half of these will not be needed by the time I'm done.

Reply:
I think `nondeterminism` was testing white space trimming options, as in different inputs could end up with the same output because of newline replacements and trimming in the parser.

Most of the others I've already chopped these fonts and more.

Reply:
Fantastic. Thank you!